cpp/src/arrow/compute/exec.cc

```diff
@@ -1123,6 +1142,34 @@ class VectorExecutor : public KernelExecutorImpl<VectorKernel> {
     }
   }
 
+  Status SetupPreallocation(const std::vector<Datum>& args) {
```

Comment:
Very hard to review this code extraction. Was this code churn necessary to achieve the goals of this PR?

Reply:
> Was this code churn necessary to achieve the goals of this PR?

Yes, that's one of the goals of this pr, we could use a more precise pre-allocation way for vector-executor.

The main change is to support the detection of null values ​​when `kernel_->null_handling == NullHandling::INTERSECTION`, so as to skip some unnecessary preallocations, including the null value detection of chunk-array, which was not done before.

Because this logic is decoupled from the scheduling of VectorExecutor Execute, it is better to consider extracting it.

> Whats the difference between it and `SetupPreallocation` in Scalar?

The difference between this function and `SetupPreallocation` of `ScalarExecutor` is that ScalarExecutor does not have an independent `chunk_exec`, so it needs to use `preallocate_contiguous_` to confirm whether it needs to execute `PrepareOutput` according to chunks through `span_iterator_`, which is not required in `VectorExecutor`.

Reply:
> The difference between this function and SetupPreallocation of ScalarExecutor is that ScalarExecutor does not have an independent chunk_exec, so it needs to use preallocate_contiguous_ to confirm whether it needs to execute PrepareOutput according to chunks through span_iterator_, which is not required in VectorExecutor.

Do you think we can extract `SetupPreallocation` to base class and extract common logic out, then implement handling for chunked array with:

```
void SetupPreallocation(...) final {
  Base::SetupPreallocation(...)
  ...
}
```

?

And what if `kernel_->null_handling == NullHandling::OUTPUT_NOT_NULL`?

Reply:
> And what if `kernel_->null_handling == NullHandling::OUTPUT_NOT_NULL`?

`Kernel output is never null and a validity bitmap does not need to be allocated`.
validity_preallocated_ just preallocate the validity bitmap-buffer which unnecessary for `OUTPUT_NOT_NULL`.